### PR TITLE
Migrate Button to Typescript and convert to function component

### PR DESCRIPTION
**Pull request by igorbrasileiro**
#### What is the purpose of this pull request?
As the title says. Related issue: #1061 

#### Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/modules/withForwardedRef.tsx [hunk 1]
```diff
@@ -1,21 +1,23 @@
...
+  ComponentType,
+} from 'react'
 import PropTypes from 'prop-types'
 
 // For more info see: https://stackoverflow.com/a/51127130/10725088
-const Element =
-  typeof window === 'undefined' || typeof window.Element === 'undefined'
-    ? function() {}
-    : window.Element
-
```
Comment (emersonlaurentino) [inline react/modules/withForwardedRef.tsx]: why did you delete that?

File: react/components/Button/index.tsx [hunk 2]
```diff
@@ -1,11 +1,124 @@
...
+  icon: false,
+  type: 'button',
+  isLoading: false,
+  isGrouped: false,
+  isFirstOfGroup: false,
+  isLastOfGroup: false,
+  isActiveOfGroup: false,
+}
+
+class Button extends Component<InferProps<typeof propTypes>> {
```
Comment (matheusps) [inline react/components/Button/index.tsx]: This could be an `FC` since you are upgrading.

Comment (igorbrasileiro) [reply react/components/Button/index.tsx]: Ok @matheusps !

File: react/modules/withForwardedRef.tsx (hunk 1 above)
Comment (klzns) [reply react/modules/withForwardedRef.tsx]: That change will break SSR

Comment (igorbrasileiro) [reply react/modules/withForwardedRef.tsx]: Yeah, I put it back!
